app/controllers/api/v2/qc_files_controller.rb

```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+module Api
+  module V2
+    # Provides a JSON API controller for QcFile
+    # See: http://jsonapi-resources.com/ for JSONAPI::Resource documentation
+    class QcFilesController < JSONAPI::ResourceController
+      # By default JSONAPI::ResourceController provides most the standard
+      # behaviour, and in many cases this file may be left empty.
+    end
+
+    class QcFileProcessor < JSONAPI::Processor
+      # Validate that attributes contains both the filename and the contents of the QcFile.
+      def validate_required_attributes(attributes)
+        errors = []
+
+        filename = attributes[:filename]
+        errors += JSONAPI::Exceptions::ParameterMissing.new('filename').errors if filename.nil?
+
+        contents = attributes[:contents]
+        errors += JSONAPI::Exceptions::ParameterMissing.new('contents').errors if contents.nil?
+
+        [filename, contents, errors]
+      end
+
+      # Create a temporary file with the contents.
+      def create_tempfile(filename, contents)
+        # The filename for a Tempfile is passed as an array with the basename and extension.
+        # e.g. [ 'file', '.csv' ]
+        filename = [File.basename(filename, '.*'), File.extname(filename)]
+        Tempfile.open(filename) do |file|
+          file.write(contents)
+          return file
+        end
+      end
+
+      # Override the default behaviour for a JSONAPI::Processor when creating a new resource.
+      # We need to parse the filename and contents attributes so that we can generate a temporary file for the new
+      # QcFile model object. Replacing the fields with these values after the new QcFile was created does not generate
+      # the file correctly in the database. The CarrierWave library needs these values sooner.
```

Comment:
I vaguely remember when you upload a file with the same name, the library is storing the file with (number) suffix. I am trying to understand if that would be an issue here. The filename in request header and filename we are dealing with might end up different if the same file is uploaded.

Reply:
I couldn't spot any problems, but then it's hard to test access by multiple people at the same time.  Creating a `TempFile` does put your file in a folder with a unique number, so I don't think any issues are going to happen.

Reply:
I meant CarrierWave opens a file for the upload and reads the content from request body into the file, so it does not read into memory. I meant that file's name might be suffixed with a number and I suspected a mismatch but as I can't remember and find the details, ignore that.